### ISSUE TITLE
no-unsafe-any: should not complain about function calls which parameters are using any

### ISSUE THREAD

**Issue by CSchulz** [palantir/tslint#2728]
### Bug Report

- __TSLint version__: 5.2.0
- __TypeScript version__: 2.2.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
// part of angular
export declare function inject(tokens: any[], fn: Function): () => any;

// my test
it('should return a new shift', inject([ShiftService, MockBackend], (service: ShiftService, backend: MockBackend) => {
```

with `tslint.json` configuration:

```json
{
  "no-unsafe-any": true,
}
```

#### Actual behavior
`Unsafe use of expression of type 'any'.`

#### Expected behavior
Should not complain about the function `inject` call.

**Comment by andy-hanson** [comment]
Could you provide a reproducible example?

**Comment by CSchulz** [comment]
Here is a repro repo 

https://github.com/CSchulz/tslint-no-unsafe-any-showcase

**Comment by ajafff** [comment]
I cloned your repo and and tried to reproduce the lint failures.

Ignoring `test.ts` I got the following:
* `app` is any at https://github.com/CSchulz/tslint-no-unsafe-any-showcase/blob/master/src/app/app.component.spec.ts#L23
* `compiled` is any at https://github.com/CSchulz/tslint-no-unsafe-any-showcase/blob/master/src/app/app.component.spec.ts#L30

These are expected failures.

Maybe there is something missing in your repro repo?

**Comment by CSchulz** [comment]
I think it is an issue of resolving dependencies. Here is my [yarn.lock](https://gist.githubusercontent.com/CSchulz/cfdafe2cd320e79d21277413241e0fcf/raw/6b6b1e156a145d97c5a619b46ea373f77fdcf519/yarn.lock) file.

I have tested it again in this moment.

**Comment by CSchulz** [comment]
I have the feeling that there are some incompatibilities between different extensions for tslint.

I have tested it again in this moment and I got the same issues. I will try to upload a shrinkwrap file to show my depdency tree.

**Comment by CSchulz** [comment]
Seems to be a collision with no-unused-variable. You should reproduce it with the following rules:

```
    "no-unsafe-any": true,
    "no-unused-variable": [
      true,
      "check-parameters",
      {
        "ignore-pattern": "^_"
      }
    ]
```

**Comment by andy-hanson** [comment]
May be #2649.

**Comment by CSchulz** [comment]
Yes could be possible, I was not sure. 

I am still wondering how I got this error with the initial ruleset.

**Comment by CSchulz** [comment]
I can avoid the issue with the initial ruleset after reinstalling the dependencies from scratch. I can't tell which dependencies make trouble with each other.

**Comment by CSchulz** [comment]
With the usage of `@Injectable()` it creates a false-positive, I have updated the repository.

**Comment by andy-hanson** [comment]
`Injectable()` is declared to be of type `any`. `@angular/core/src/di/metadata.d.ts` defines `export declare const Injectable: InjectableDecorator;` and `InjectableDecorator` declares `(): any;`.
Or you could just hover over the `@Injectable()` in the IDE and saw it show up as `(alias) Injectable(): any`.

**Comment by CSchulz** [comment]
So every type which is finally any leads to this violation?

**Comment by andy-hanson** [comment]
Not sure what you mean -- only `any` is `any`. Do you mean that `Injectable` is `() => any` and not `any`? But the error isn't for `Injectable`, it's for `Injectable()`, which is just `any`.

**Comment by CSchulz** [comment]
I meant everytime any is used somewhere in the declaration like in this case `() => any`, it triggers a violation?

How can you prevent it? Using some type assertion using with the `Injectable()` decorator?

**Comment by andy-hanson** [comment]
You could always help improve the typings and get rid of `any`. The rule will also ignore anything explicitly cast with `x as any`.

**Comment by CSchulz** [comment]
Would it be possible to define something better f.e. for the Injectable `() => Function`?

**Comment by nicolashenry** [comment]
@CSchulz I use this in a .d.ts file if it can help:

```typescript
import core from '@angular/core';

declare module '@angular/core' {
	export interface InjectableDecorator {
		(): ClassDecorator;
	}
	export const Injectable: InjectableDecorator;

	export interface HostBindingDecorator {
		(hostPropertyName?: string): PropertyDecorator;
	}
	export const HostBinding: HostBindingDecorator;

	export interface HostListenerDecorator {
		(eventName: string, args?: string[]): MethodDecorator;
	}
	export const HostListener: HostListenerDecorator;
}
```

**Comment by CSchulz** [comment]
@nicolashenry Thanks, did you consider to contribute it to angular itself?

**Comment by nicolashenry** [comment]
@CSchulz I will but it will take some time.

**Comment by CSchulz** [comment]
Thanks, do you have some typings also for the provider definitions?

**Comment by CSchulz** [comment]
Do I have to name the file special or do I need a different configuration for tslint?

I have tried to configure the folder as types root in the tsconfig.json without any change.

**Comment by OliverJAsh** [comment]
I'm also getting extraneous errors when using this rule in combination with no-unused-variables. If I disable that rule, the errors for no-unsafe-any go. 

I'm also seeing extraneous errors with no-inferred-object when used in combination with no-unused-variables

**Comment by miltonhowe** [comment]
Has anyone found a solution for this? I tried the `d.ts` workaround suggested by @nicolashenry but that caused cascading problems. I am using Angular 5.0.0 and want to implement `no-unsafe-any` but it is showing up on every `@injectable` decorator.

**Comment by adidahiya** [comment]
Going to close this in favor of #2649, as that seems to be the root of these weird side-effectful issues around no-unused-variable.

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
